project-docs/javascript-linter-spec.md

```diff
@@ -0,0 +1,163 @@
+
+## JavaScript recipes
+
+We've defined the following recipes for the JS docs:
+
+- [javascript-class](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-class.yaml)
+- [javascript-namespace](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-namespace.yaml)
+- [javascript-constructor](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-constructor.yaml)
+- [javascript-method](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-method.yaml)
+- [javascript-property](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-property.yaml)
+- [javascript-error](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-error.yaml)
+- [javascript-language-feature](https://github.com/mdn/stumptown-content/blob/master/recipes/javascript-language-feature.yaml)
+
+## JavaScript recipe ingredients
+
+In total these recipes use the following ingredients:
+
+prose ingredients:
+
+- prose.short_description
+- prose.description?
+- prose.see_also
+- prose.syntax
+- prose.message
+- prose.error_type
+- prose.what_went_wrong
+- prose.\*
+
+data ingredients:
+
+- data.interactive_example?
+- data.class_constructor
+- data.constructor_properties?
+- data.static_properties?
+- data.static_methods?
+- data.instance_properties?
+- data.instance_methods?
+- data.examples
+- data.specifications
+- data.browser_compatibility
+
+All ingredients except two are demarcated in the page using an `<h2>` heading. The exceptions are `prose.short_description` and `data.interactive_example`.
+
+### Prose ingredients
+
+With three exceptions, all prose ingredients impose the same requirement on the page: it must have an `<h2>` section whose ID attribute maps to the named section. The mapping is simply that `prose.` is removed, and the first letter is capitalized:
+
+```
+prose.syntax      ->    Syntax
+prose.see_also    ->    See_also
+```
+
+The exceptions are `prose.short_description`, `prose.description?`, and `prose.*`.
+
+#### prose.short_description
+
+This ingredient isn't demarcated by an `<h2>` at all. To satisfy the ingredient, a page must:
+
+- contain a `<p>` element
+- that is not a note or a warning
+- before the interactive example or the first `<h2>`, whichever comes first
+
+#### prose.description?
+
+This is an optional prose ingredient. To satisfy the ingredient, if the page contains an `h2#Description`, then it must appear in the place specified by the recipe.
+
+#### prose.\*
+
+This ingredient enables writers to include extra sections, but ensuring that they always appear in the same place in the page. To satisfy this ingredient: if the page contains any `<h2>` headings that are not expected by the recipe, then they must appear at the point in the page that the `prose.*` ingredient is listed in the recipe.
```

Comment:
> any `<h2>` headings that are not expected by the recipe

Interesting! I had previously understood `prose.*` to mean: any `<h2>`-demarcated sections that are not otherwise _specified_. I think your definition may might have implications for _optional_ data ingredients—though I guess that depends on the earlier optionality question. This is complicated and I'm not sure if it matters yet, but I thought I should write it down. 😆

Reply:
I don't follow. Can you elaborate on the difference between what you thought before and what you think this is saying? It's just as possible that I'm expressing it wrong here (and very likely that this isn't clearly enough expressed, since we have doubt about what it means).

Reply:
OK, now that I've written this out, I'm less worried about this thing. But here's what I was thinking:

I thought `prose.*` would be any section that would not otherwise appear as part of a recognized ingredient. For example, suppose you had a page with a "Specifications" section with the proper macro call, but that page's recipe did not include `data.specifications`. It seems to me that we could reasonably warn about that because the page doesn't call for that ingredient.

But `prose.*` complicates matters. Under your definition, the presence of `prose.*` in the recipe might allow for content that matches the `data.specifications` ingredient, even if that ingredient is not part of the recipe. In other words, there's no way to _forbid_ `data.specifications` content from appearing in a page, if it's in a range of sections consumed by `prose.*`.

In my definition, `prose.*` could _never_ consume a "Specifications" H2; a "Specifications" H2 could _only_ be used to mark the start of content that adheres to `data.specifications`. Now that I've spelled this out, it's probably a bad idea because you'd have to contort yourself (e.g., writing "Notes on specifications") to avoid clashing with section headings that are already "taken" by an ingredient.

Reply:
In practice I can't think of a place we're likely to want a page to use a heading that's not in its recipe but is in another recipe. But it doesn't seem like a thing we want to exclude.

I suppose one factor here is: the more heterogeneous are the different doc types that we expect this system to govern, then the less sense it seems to make for one recipe to be able to prevent another recipe from using a particular H2.

Reply:
Hmm. Maybe the right thing to do here is experiment with the more restrictive option. That would show us whether (and how) it breaks for the JS docs (I don't think this is likely, because they're so alike). When we expand to other areas, we may find that `prose.*` needs to be redefined, or that we need to introduce another, more permissive ingredient type.

Reply:
I don't think it's likely to break for the JS docs either, so this has no real practical effect. But it just seems wrong for recipe A to have any effect on what recipe B can include.

Another way to think of it: at the moment there are two categories of prose H2 headings:

* ones that are explicitly defined in its recipe: it must have these (or may have them if the ingredient is optional), and they must be in the defined place
* ones that are not explicitly defined in its recipe, which the page is allowed if `prose.*` is present

This adds a third category:
* ones it's just not allowed to use, regardless of `prose.*`, because an unrelated recipe is using them over in the SVG docs or something

I don't see the rationale for this category. It makes the rules more complex, and means authors have to know about all the recipes to write correct docs, rather than just their own (it makes recipes correctness dependent on other recipes).

But, like I say, I think this has no practical effect on the JS docs either way, so I will make this change if you're still not convinced.

Reply:
Yeah, it's fine to leave this as is. I don't think this obligates authors to know about all the recipes, but it would obligate them to know about all the possible ingredients. That's still kinda annoying.

But it's also not essential: just because `prose.*` is specified permissively, doesn't mean we can't have additional linting that finds possibly inadvisable content. Here's a (somewhat contrived) example: imagine you've got a JS namespace page that accidentally contains a class constructor section. Depending on where that class constructor section appears in the page, it can be a valid page, even though the content is garbage no matter its position. But there's nothing stopping us from warning for unexpected ingredients, even if the spec says it's ultimately valid.